src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -2832,9 +2837,26 @@ ErrorIfContainsUnsupportedSubquery(MultiNode *logicalPlanNode,
 	/* currently in the planner we only allow one subquery in from-clause*/
 	Assert(list_length(subqueryMultiTableList) == 1);
 
-	restrictionEquivalenceForPartitionKeys =
-		RestrictionEquivalenceForPartitionKeys(plannerRestrictionContext);
-	if (!restrictionEquivalenceForPartitionKeys)
+	/*
+	 * We're checking two things here:
+	 *    (i)   If the query contains a top level union, ensure that all leaves
+	 *          return the partition key at the same position
+	 *    (ii)  Else, check whether all relations joined on the partition key or not
+	 */
+	if (ContainsUnionSubquery(originalQuery))
+	{
+		if (!SafeToPushdownUnionSubquery(relationRestrictionContext))
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot pushdown the subquery since all leaves of "
+								   "the UNION does not include partition key at the "
+								   "same position"),
+							errdetail("Each leaf query of the UNION should return "
+									  "partition key at the same position on its "
+									  "target list.")));
+		}
+	}
+	else if (!RestrictionEquivalenceForPartitionKeys(plannerRestrictionContext))
```

Comment:
Does this mean we do not need to check`RestrictionEquivalenceForPartitionKeys` if the query contains top level union ?